NuRadioReco/framework/parameter_storage.py

```diff
@@ -0,0 +1,57 @@
+import logging
+logger = logging.getLogger('NuRadioReco.framework.parameter_storage')
+
+
+
+def check_key(func):
+
+    def wrapper(*args):
+        key = args[1]
+        parameter_class = args[0]._parameter_type
+        if not isinstance(key, parameter_class):
+            logger.error(f"Parameter {key} needs to be of type {parameter_class}")
+            raise ValueError(f"Parameter {key} needs to be of type {parameter_class}")
+        return func(*args)
+
+    return wrapper
+
+
+class ParameterStorage:
+
+    def __init__(self, parameter_type):
+        self._parameters = {}
+        self._parameter_covariances = {}
+        self._parameter_type = parameter_type
+
+    def __setitem__(self, key, value):
+        self.set_parameter(key, value)
+
+    def __getitem__(self, key):
+        return self.get_parameter(key)
+
+    @check_key
+    def get_parameter(self, key):
+        return self._parameters[key]
```

Comment:
why not instead of the decorator just have a method
```
def __check_key(self, key):
    if not isinstance(key, self._parameter_type):
    ...
```
and then in e.g. get_parameter
```
def get_parameter(self, key):
    self.__check_key(key)
    return self._parameters[key]
```
?
The advantage being you don't have to mess around with `args` and `kwargs` in the decorator.